src/EditorFeatures/Core/LanguageServer/Handlers/Completion/CompletionHandler.cs

```diff
@@ -169,101 +185,85 @@ bool IsValidTriggerCharacterForDocument(Document document, char triggerCharacter
-                    completionItem.TextEdit = textEdit;
-                }
-                // If the feature flag is off, return an InsertText.
                 else
                 {
-                    completionItem.InsertText = SymbolCompletionItem.TryGetInsertionText(item, out var insertionText) ? insertionText : completeDisplayText;
-                }
-
-                var commitCharacters = GetCommitCharacters(item, commitCharacterRulesCache, supportsVSExtensions);
-                if (commitCharacters != null)
-                {
-                    completionItem.CommitCharacters = commitCharacters;
-                }
-
-                if (completionItem is LSP.VSInternalCompletionItem vsCompletionItem)
-                {
-                    vsCompletionItem.Icon = new ImageElement(item.Tags.GetFirstGlyph().GetImageId());
-                }
-
-                return completionItem;
-
-                static async Task<LSP.TextEdit> GenerateTextEdit(
-                    Document document,
-                    CompletionItem item,
-                    CompletionService completionService,
-                    SourceText? documentText,
-                    TextSpan? defaultSpan,
-                    LSP.Range? defaultRange,
-                    CancellationToken cancellationToken)
-                {
-                    Contract.ThrowIfNull(documentText);
-                    Contract.ThrowIfNull(defaultSpan);
-                    Contract.ThrowIfNull(defaultRange);
-
                     var completionChange = await completionService.GetChangeAsync(
-                        document, item, cancellationToken: cancellationToken).ConfigureAwait(false);
-                    var completionChangeSpan = completionChange.TextChange.Span;
+                        document, roslynItem, cancellationToken: cancellationToken).ConfigureAwait(false);
 
-                    var textEdit = new LSP.TextEdit()
+                    if (completionChange.TextChanges.Length > 1 && lspItem is LSP.VSInternalCompletionItem vsCompletionItem)
+                    {
+                        // We have more than one text change.  This means the edit is broken up into a main edit around the
+                        // cursor location and additional edits.
+                        // Due to https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1513155 in VS we cannot provide the additional text edits
+                        // on resolve only as the client may not call us.  Instead we provide nothing and force the client to resolve us.
+                        // This is not ideal for filtering as we have to rely on the client inferred range, but is currently the best we can do.
+                        //
+                        // We also cannot provide the additional edits and the main text edit upfront as Razor currently cannot performantly
```

Comment:
To be honest, I'm not sure we'd want to anyways. Additional Text Edits for every item is expensive serialization wise 😄